src/core/Akka/Actor/Futures.cs

```diff
@@ -117,66 +117,44 @@ public static Task<T> Ask<T>(this ICanTell self, object message, TimeSpan? timeo
         /// This exception is thrown if the system can't resolve the target provider.
         /// </exception>
         /// <returns>TBD</returns>
-        public static async Task<T> Ask<T>(this ICanTell self, Func<IActorRef,object> messageFactory, TimeSpan? timeout, CancellationToken cancellationToken)
+        public static async Task<T> Ask<T>(this ICanTell self, Func<IActorRef, object> messageFactory, TimeSpan? timeout, CancellationToken cancellationToken)
         {
             await SynchronizationContextManager.RemoveContext;
 
             IActorRefProvider provider = ResolveProvider(self);
             if (provider == null)
                 throw new ArgumentException("Unable to resolve the target Provider", nameof(self));
 
-            return (T)await Ask(self, messageFactory, provider, timeout, cancellationToken);
-        }
-
-        /// <summary>
-        /// Resolves <see cref="IActorRefProvider"/> for Ask pattern
-        /// </summary>
-        /// <param name="self">Reference to someone we are sending Ask request to</param>
-        /// <returns>Provider used for Ask pattern implementation</returns>
-        internal static IActorRefProvider ResolveProvider(ICanTell self)
-        {
-            if (self is ActorSelection)
-                return ResolveProvider(self.AsInstanceOf<ActorSelection>().Anchor);
-
-            if (self is IInternalActorRef)
-                return self.AsInstanceOf<IInternalActorRef>().Provider;
-            
-            if (ActorCell.Current != null)
-                return InternalCurrentActorCellKeeper.Current.SystemImpl.Provider;
-
-            return null;
-        }
-        
-        private static async Task<object> Ask(ICanTell self, Func<IActorRef, object> messageFactory, IActorRefProvider provider,
-            TimeSpan? timeout, CancellationToken cancellationToken)
-        {
-            TaskCompletionSource<object> result = TaskEx.NonBlockingTaskCompletionSource<object>();
+            var result = TaskEx.NonBlockingTaskCompletionSource<T>();
 
             CancellationTokenSource timeoutCancellation = null;
             timeout = timeout ?? provider.Settings.AskTimeout;
-            var ctrList = new List<CancellationTokenRegistration>(2);
```

Comment:
Removed `List<T>` allocation and replaced with two `Nulalble<CancellationTokenRegistration>`